### ISSUE TITLE
Enable/disable loop points should be more obvious

### ISSUE THREAD

**Issue by tresf** [LMMS/lmms#3069]
The enable/disable loop points button is a bit backwards.  It's gray-ed out when it's inactive and just "regular" when it's active.

At a glance, this makes it hard to know when the loop points are enabled:

![screen shot 2016-10-08 at 11 53 37 am](https://cloud.githubusercontent.com/assets/6345473/19214320/05e56668-8d4e-11e6-9c65-ea34c7699141.png)

Gray-ed out from a UI perspective generally means it's not accessible.  I think instead, the loop points should behave more like an LED toggle and be very obvious when enabled.

![screen shot 2016-10-08 at 11 53 37 am](https://cloud.githubusercontent.com/assets/6345473/19214342/8f872262-8d4e-11e6-95a2-3c2abf47e3d0.png)

This way when a composer finds their project arbitrarily jumping away from the cursor area, they have a visual indicator as to why.

**Comment by BaraMGB** [comment]
This looks a little bit like a  overkill to me. Perhaps we can make it like the pencil button. When it's active it looks pushed in.

**Comment by tresf** [comment]
> This looks a little bit like a overkill to me. Perhaps we can make it like the pencil button. When it's active it looks pushed in.

Pencil is a radio button, not an on/off toggle.  Also, Pencil has a persistent cursor which illustrates it's action to the user.  Looping should be very obvious to the end-user.  Furthermore, suppressed... does this mean looping is enabled or disabled?  If your mind needs to think for 0.5 seconds about it to use it, it's bad UI.

**Comment by RebeccaDeField** [comment]
Does anyone know how other DAWs handle the enable/disable loop points function?

**Comment by SirBothersome** [comment]
Of my brief use of FL Studio, not much different than this AFAIK

**Comment by tresf** [comment]
> Does anyone know how other DAWs handle the enable/disable loop points function?

 * Ableton and Pro Tools will highlight the entire loop region (patterns too) and use a separate button for playing the loop. [sample](https://cloud.githubusercontent.com/assets/6345473/19220589/d8b2bd88-8dfe-11e6-98eb-09a3f66f1c36.png)

 * Audacity (not a DAW) just [loops what's selected](https://cloud.githubusercontent.com/assets/6345473/19220609/3e8cface-8dff-11e6-89e7-a35c85c55f4d.png), assuming that's what you wanted.

 * Logic Pro uses a very, [very distinguishable loop area](https://cloud.githubusercontent.com/assets/6345473/19220622/8a6971a2-8dff-11e6-9734-98fc749fb698.png) but I've no idea how it's toggled.

 * Bitwig studio appears to use a [bright button](https://cloud.githubusercontent.com/assets/6345473/19220640/d4202c28-8dff-11e6-8092-94792f445e8c.png) although others that have used it would need to confirm.

**Comment by mikobuntu** [comment]
> Bitwig studio appears to use a bright button although others that have used it would need to confirm.

* All Bitwig buttons when not in use are grey with a white logo and when in use ( toggled ) are orange with a black logo.

* Ardour5 : when buttons are not in use are grey with white logos ( small black outline ) and when in use ( toggled ) only the grey changes to green. Also the loop region on the song editor is all highlighted with a green transparency. I find this behaviour the most obvious from most DAWS that I use.

* QTractor : Is probably the worst to see what is going on regarding loop area. Hovering over the loop button uses the same image/colour/css/whatever as actually enabling loop play.

*Hydrogen ( drum machine ) Just changes the button colour from grey to blue on toggle.

**Comment by Umcaruje** [comment]
I think also that a nicer icon should be on the button. The current icon made sense back when we had the loop markers, but now its just confusing. Most DAW's have an icon similar to this: ![screenshot from 2016-10-09 18 51 20](https://cloud.githubusercontent.com/assets/6282045/19222121/7254655c-8e51-11e6-99e0-a32c09cd7c6e.png) a simple line that clearly indicates looping. 

I also like @BaraMGB's idea of making all the toggle buttons get creased in as pressed, like we do with the radio buttons.

**Comment by RebeccaDeField** [comment]
I agree with @Umcaruje about changing the icon.

I would like to propose that we start with changing the button to be pressed so that it matches the other behavior in the program and then if users do not find that to be distinguishable enough, we could use color instead.

I can tackle the new icon, but someone more familiar with the code side of things would need to change the loop button styling.

**Comment by tresf** [comment]
Rough mockup... 

I agree this is better visually although now we're starting to introduce some inconsistencies in our GUI.  (play toggle, auto-scrolling toggle, etc).  If we are to keep this consistent, I'd like auto-scroll to be default and non-autoscroll to be the supressed button.  Play should be supressed while playing, etc.

![image](https://cloud.githubusercontent.com/assets/6345473/19275889/042d8ad8-8fa3-11e6-8580-fc5e237529c1.png)

**Comment by RebeccaDeField** [comment]
Idea for loop points icon:
![loop_points_on](https://cloud.githubusercontent.com/assets/7960169/19501438/91b19db0-955a-11e6-898b-b30636e174bf.png)

**Comment by RebeccaDeField** [comment]
After some discussion on Discord, this is what the icon looks like:
![loop_points_on 1](https://cloud.githubusercontent.com/assets/7960169/19615778/8df0a37c-97b9-11e6-93ec-2d20b8b8259e.png)

![loop_points_on](https://cloud.githubusercontent.com/assets/7960169/19615803/499e7f5e-97ba-11e6-89ac-8210f67b30eb.png)
^^^ Transparent .png here

Now that we have a new icon to work with, is there someone that would like to code up @tresf's [proposal for the pressed button](https://github.com/LMMS/lmms/issues/3069#issuecomment-252961244)?

**Comment by smithjessk** [comment]
@RebeccaDeField I am interested!

**Comment by Umcaruje** [comment]
Could we just make the icon green when its enabled and do the same with the autoscroll button? I think it would fit.

**Comment by tresf** [comment]
Auto scroll is on by default.  Would that be confusing from a UI perspective to have it bright green on project load?

**Comment by Spekular** [comment]
@tresf I think so.

**Comment by RebeccaDeField** [comment]
> Would that be confusing from a UI perspective to have it bright green on project load?

@tresf I think so as well. I've came up with a few alternate ideas for this, but none of them made sense to me from a UI consistency standpoint when tested in the program.

Anyone have any other ideas?

**Comment by tresf** [comment]
I think consensus is the combination of a few posts above:

 * Make the color different per: @mikobuntu
   * All Bitwig buttons when not in use are grey with a white logo and when in use ( toggled ) are orange with a black logo. [ this is a very good way of showing what is going on ] [[ imo :) ]]
 * Improve the loop logo per @Umcaruje:
   * I think also that a nicer icon should be on the button. The current icon made sense back when we had the loop markers, but now its just confusing. 

If the only way to make this consistent is to also do this to the pencil, so be it.

**Comment by tresf** [comment]
> If the only way to make this consistent is to also do this to the pencil (play, etc) in the same fashion, so be it.

Going out on a limb... One other idea from a UI perspective is to make two tones of green as to illustrate severity.  I'll call these tones "Meh" and "OMG".  Stuff like Loop and Play can use "OMG" where pencil and section can use "Meh".  That way we can maintain a consistent UI theme without sacrificing usability.

**Comment by RebeccaDeField** [comment]
> Improve the loop logo per @Umcaruje:

loop_points_on 1

**Comment by mikobuntu** [comment]
Here's how I envisage my Loop_Region_Overlay to look in LMMS.
![lmmsloopoverlay](https://cloud.githubusercontent.com/assets/335217/20784899/0dc43366-b795-11e6-938f-c375f81cec23.png)

Comparing to Ardours view of Loop_Region.
![ardourloopoverlay](https://cloud.githubusercontent.com/assets/335217/20784921/30a58e16-b795-11e6-9b6b-a67874f14d5c.png)

**Comment by musikBear** [comment]
@mikobuntu it looks good, but only ok If the transparency is cpu-lean, imo

**Comment by tresf** [comment]
> but only ok If the transparency is cpu-lean, imo

Please stop making these statements.  No one wants the software to run slowly and blindly mentioning it on every UI change is obvious and superfluous.

The draw routine for the piano roll should be able to handle this with little or additional overhead.  The codebase is shared with this logic/change https://github.com/LMMS/lmms/issues/3019#issuecomment-260156089.

**Comment by RebeccaDeField** [comment]
@mikobuntu I would turn down the opacity a touch, but think your idea makes the loop point a lot more obvious.

I noticed in your screenshot of Ardours the white audio graph and midi track notes are above the transparent layer which I believe is done so that the producer can still easily identify those elements. Does anyone know if that would be possible for us to implement as well?

**Comment by tresf** [comment]
> Does anyone know if that would be possible for us to implement as well?

Theoretically, anything is possible, but I believe the highlighted region is a topic for another issue, unless of course @mikobuntu plans to work on this. 🏖

**Comment by mikobuntu** [comment]
@RebeccaDeField Yes I think the opacity may have been a bit higher than needed, this was just a mockup to show roughly how my idea would look. I will leave the colour stuff up to the experts like yourself :) @tresf  I'm unsure if I would have enough knowledge to be able to work on this to be honest tho. I could probably have a look through the QT docs for some info regarding overlays.

**Comment by Umcaruje** [comment]
I like @mikobuntu's idea and I would be willing to work on it, but I'm in middle of my semester so I don't have that much free time on my hands. So maybe when things cool down I can take this on and finish that piano roll grid

**Comment by tresf** [comment]
@mikobuntu if you have time to start this, @Umcaruje's PR is a good place to start looking at the code:

https://github.com/LMMS/lmms/pull/3062/files

The paintEvent stuff is about as straight forward as it gets.  Unfortunately the routines are huge, so finding out where best to put the paint code can get tricky. 👍

**Comment by RebeccaDeField** [comment]
_And then out of the blue, a wild idea attacks.._

![idea](https://cloud.githubusercontent.com/assets/7960169/20812890/3332f33e-b7c8-11e6-9cf7-e2a5c6402b0a.png)

This would be consistent with the UI because:
![buttons](https://cloud.githubusercontent.com/assets/7960169/20813042/f18e1570-b7c8-11e6-8062-7c2028348dbd.png)

**Comment by tresf** [comment]
Valiant idea.  From a UI perspective, I'd say the green is less bold than white, so it actually has the opposite effect. 😕

**Comment by RebeccaDeField** [comment]
@tresf Ok, then maybe let's just code up the green button idea and test that out. It seems to be the best option so far.

**Comment by tresf** [comment]
>  just code up the green button idea and test that out. It seems to be the best option so far.

The loop points button is of type `NStateButton`, which has some GUI code regarding state here:

https://github.com/LMMS/lmms/blob/master/src/gui/widgets/NStateButton.cpp#L81

Since `NStateButton` is custom toggle, I don't think we can use CSS "pressed" indicator, so perhaps a start would be to hard-code a background color (untested):

```diff
  setIcon( m_states[m_curState].first );
+ setBackground(n > 0 ? Qt::green : Qt::transparent);
  emit changedState( m_curState );
```

If this works, I think we would need to use a custom `qproperty-...`.  This isn't my area of expertise, but hopefully that's a start.

**Comment by blurrred** [comment]
I have to agree with this, and thank the user who submitted it. I could not figure out how to get a repeat to work until I came across this issue.

**Comment by RebeccaDeField** [comment]
Since changing the button to be green seems to be at a standstill, I'm thinking about changing active icons to be green. It solves the immediate issue of confusion which is making some newer icons already outside of the palette and style.

If I don't hear any objections about this, I will make a PR for it.

**Comment by musikBear** [comment]
@RebeccaDeField Good idea, except maby for this 
![image](https://user-images.githubusercontent.com/6662066/87879800-2f868a80-c9ed-11ea-92f4-64e0f9e45cc0.png), because it is always selected in one of its states. Else all  should  be green i guess..

**Comment by RoxasKH** [comment]
I know probably importante decision had already been made. But anyway i think in the future LMMS should get an highlighted background for that button whn is active, something like this (i know it's horrible it's just to give an understanding, so like pressed but with a green background)

![Senzanome](https://user-images.githubusercontent.com/40141826/87880255-948faf80-c9f0-11ea-884e-5d19bd998895.png)

This because it's the most common behaviour in DAWs.

Here's some examples

FL Studio 12: ![flloop](https://user-images.githubusercontent.com/40141826/87880269-bb4de600-c9f0-11ea-80ce-d3dff73f2950.PNG)
Bitwig: ![bitwigloop](https://user-images.githubusercontent.com/40141826/87880273-c43eb780-c9f0-11ea-99ee-bfac3b9a0479.png)
Ableton: ![abletonlloop](https://user-images.githubusercontent.com/40141826/87880276-c9036b80-c9f0-11ea-9a44-d3137e43f013.PNG)
Reason: ![reasonloop](https://user-images.githubusercontent.com/40141826/87880282-d28cd380-c9f0-11ea-82e7-85785ced8233.PNG)
Cubase: ![cubaseloop](https://user-images.githubusercontent.com/40141826/87880286-dae50e80-c9f0-11ea-9cf7-09315efaca42.PNG)
Ardour: ![ardourloop](https://user-images.githubusercontent.com/40141826/87880290-e20c1c80-c9f0-11ea-87b4-47040171d99e.PNG)

Even if Budislav concept shows an FL20 like behaviour, so maybe that'll be like that.

Budislav: ![immagine](https://user-images.githubusercontent.com/40141826/87880323-213a6d80-c9f1-11ea-978e-e3344a46442f.png)
FL20: ![immagine](https://user-images.githubusercontent.com/40141826/87880332-2f888980-c9f1-11ea-812b-e120dc2473c7.png)

**Comment by RebeccaDeField** [comment]
@RoxasKH I agree, this would be ideal. The coding work behind implementing that seemed to be at a standstill so I was looking for a workaround that I could make happen. If we could implement it this way, we already have the improved icon ready.

An older mockup we had settled on some time ago:
![8f872262-8d4e-11e6-95a2-3c2abf47e3d0](https://user-images.githubusercontent.com/7960169/87897734-1687d480-ca01-11ea-8adf-1c471d2f6f5b.png)

A possible good addition with some design tweaking:
![0dc43366-b795-11e6-938f-c375f81cec23](https://user-images.githubusercontent.com/7960169/87897797-47680980-ca01-11ea-9775-43313c26b2e9.png)

**Comment by RebeccaDeField** [comment]
As a follow up, if someone is able to team up with me to work on this, I can make improved mockups for it.

**Comment by russiankumar** [comment]
Not a designer, but I can help in coding if required.

**Comment by RoxasKH** [comment]
> 
> 
> @RoxasKH I agree, this would be ideal. The coding work behind implementing that seemed to be at a standstill so I was looking for a workaround that I could make happen. If we could implement it this way, we already have the improved icon ready.
> 
> An older mockup we had settled on some time ago:
> 
> A possible good addition with some design tweaking:


That green gradient is kinda weird, maybe it shuld be lighter, but i got it. I love the transaprency of the second mockup.

The problem is that that button can't be styled in the css because all buttons use the same class as i experienced when i was making my theme.
If we can get a separate class for that button that could be implemented through css and be also themeable. But maybe is a bad workaround, let's see what developers say.

**Comment by tresf** [comment]
LcdWidget uses a style parameter to change its behavior.

https://github.com/LMMS/lmms/blob/174630087ef24899dfb52ef2e892c68a7fd7caf2/src/gui/widgets/LcdWidget.cpp#L255

This isn't the same thing as overloading the CSS, but might be helpful.

Knob does something closer to what's being described, it makes a special class for implementation-specific version:

https://github.com/LMMS/lmms/blob/ce5c74432e83f0eaa617cc26332f4209bc4efc7f/plugins/kicker/kicker.cpp#L230

Last, and perhaps the simplest -- is to allow multi-state buttons to be an alternate button when they're non-default.  This would encourage consistency (such as the example @musikBear makes with the play head behavior) so that it can benefit from the same "I'm obviously toggled right now" visual appearance.

**Comment by RebeccaDeField** [comment]
> Not a designer, but I can help in coding if required.

I'll take care of the design refinements. 

> That green gradient is kinda weird, maybe it should be lighter, but i got it. I love the transparency of the second mockup.

Yeah, I agree that the old mockup needs some revisiting and rethinking.

**Comment by tresf** [comment]
> That green gradient is kinda weird
> > that the old mockup needs some revisiting

Also will need to be tested against the selection-rectangle.

![image](https://user-images.githubusercontent.com/6345473/87952334-b2ecbe00-ca77-11ea-82d3-fe1f30579f5d.png)

**Comment by zacius** [comment]
FWIW i came to this thread to find out how to turn it off. 

+1 unintuitive UI 

But again, my deepest gratitude, this software has enabled me to express myself and feel a creative connection to the sounds of the world around me.

**Comment by Rossmaxx** [comment]
At current master, looks like the icon has changed and setting loop on seems to have become more obvious. If anyone disagrees, we can reopen.

**Comment by musikBear** [comment]
> At current master, looks like the icon has changed and setting loop on seems to have become more obvious. If anyone disagrees, we can reopen.

Can you screen-shoot the on/off behaviour of Master?

**Comment by Rossmaxx** [comment]
![IMG_20240615_211033](https://github.com/LMMS/lmms/assets/74815851/088cbf3b-393c-4130-9fc5-521198eaae18)

Here's loop marker on 

![Screenshot_20240615_210940](https://github.com/LMMS/lmms/assets/74815851/f2f17e16-bab4-4e6f-b8f8-38c951a3d66a)

And here's loop marker off. There is change from what is defined in the PR and it is obvious for regular users. The green is missing. If the green background should be added, just lmk and I'll reopen this.

**Comment by Rossmaxx** [comment]
#5588 was the fix.

**Comment by Rossmaxx** [comment]
Got a hint from discord that the icon change is not the fix. The button should still be highlighted. Reopening.

**Comment by musikBear** [comment]
>  The button should still be **highlighted**. 

Agree.

**Comment by Rossmaxx** [comment]
@zonkmachine how viable is it to add this change too to #7314

If you can't, I'll try something but I'll have to do some learning before i land the fix.

**Comment by zonkmachine** [comment]
> @zonkmachine how viable is it to add this change too to #7314

I'll look into it.

**Comment by zonkmachine** [comment]
![looppoints](https://github.com/LMMS/lmms/assets/6368949/f803001e-222e-43dc-a074-789c436fb39f)

Something like this? It's just a matter of editing the colors of `loop_points_on.png`

**Comment by Rossmaxx** [comment]
Nope, more like changing the background colour to highlight to green

**Comment by zonkmachine** [comment]
> Nope, more like changing the button's background colour to highlight to green

Similar how the modes draw/knife/edit in the editor change background when selected? Darker on the Default theme and lighter on Classic.

![actions](https://github.com/LMMS/lmms/assets/6368949/1426dc67-142c-46d8-a83f-4ce9a5446bdb)

**Comment by Rossmaxx** [comment]
Similar but the active background should be green

**Comment by zonkmachine** [comment]
> Similar but the active background should be green like in the linked comment above.

Yes, I occasionally do green. I don't get where that background is set in the case of the edit buttons though?

**Comment by Rossmaxx** [comment]
It's currently not available for theming. https://github.com/LMMS/lmms/issues/3069#issuecomment-264476113 in combination with a theme entry should do the trick.
